### PR TITLE
Feature add last_operation to time based cover

### LINKED ISSUE #4252: Checking the state of current_operation for time-based cover doesn't work
### The problem

I'm trying to create a time-based cover and then do some rudimentary checking of the cover's operation in a stop script, so that I can either send the reverse command when it's moving or use a double click when the cover is already stopped to tell the controller to send it to a pre-set position.

Complete yaml below, here are the relevant bits.

 I've set up a time-based cover:

```
cover:
  - platform: time_based
    id: tb_cover
    name: "LR Side Cover"
    has_built_in_endstop: true
    open_duration: 20s    
    close_duration: 10s

    open_action:
      - button.press: button_single_up

    close_action:
      - button.press: button_single_down

    stop_action:
      - script.execute: stop
```
and the appropriate script to handle the stop action:

```
script: 
 - id: stop
    then:
      - if:
          condition:
            lambda: |-
              if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_IDLE) {
                return true;
              } else if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_OPENING) {
                return false;
              } else if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_CLOSING) {
                return false;
              }            
          then:
            - button.press: button_double_down
          else:
            - button.press: button_single_down
```
The logic obviously isn't complete however I'm stuck already at distinguishing between any of the potential current_operation states, the code above (taken directly from ESPHome docs) always returns true and therefore always executes the double_down button.

Here's what that looks like in logs:

```
[20:01:08][D][cover:170]: 'LR Side Cover' - Publishing:
[20:01:08][D][cover:173]:   Position: 25%
[20:01:08][D][cover:186]:   Current Operation: OPENING
[20:01:08][D][switch:017]: 'ESP Cover LR Side Up' Turning OFF.
[20:01:08][D][switch:056]: 'ESP Cover LR Side Up': Sending state OFF
[20:01:09][D][cover:170]: 'LR Side Cover' - Publishing:
[20:01:09][D][cover:173]:   Position: 30%
[20:01:09][D][cover:186]:   Current Operation: OPENING   <------- Cover is opening
[20:01:10][D][cover:076]: 'LR Side Cover' - Setting
[20:01:10][D][cover:080]:   Command: STOP                      <------- Pushed STOP in the web controls
[20:01:10][D][button:013]: 'ESP Cover LR Side Double Down' Pressed.        <----------- Double click is pressed - should be single
```

It's entirely possible I'm missing something completely obvious... 

Thanks!

### Which version of ESPHome has the issue?

2023.2.4

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32

### Board

_No response_

### Component causing the issue

_No response_

### Example YAML snippet

```yaml
substitutions:
  esp_name: ESP Cover LR Side #Device Name
  esp_hostname: esp-cover-lr-side
  push_delay: 80ms
  ip_address: 192.168.5.105
  ip_gw: 192.168.5.1
  ip_netmask: 255.255.255.0


esphome:
  name: ${esp_hostname}

esp32:
  board: esp32dev
  framework:
    type: arduino

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  manual_ip:
    static_ip: ${ip_address}
    gateway: ${ip_gw}
    subnet: ${ip_netmask}
  ap:
    ssid: "${esp_name} Hotspot"
    password: !secret ap_password

ota:
  password: !secret ota_password

# Enable logging
logger:

# Enable Home Assistant API
api:
  encryption:
    key: !secret api_key

captive_portal:

web_server:
  port: 80
  include_internal: true 

switch:
  - platform: gpio
    id: switch_up
    pin: 25
    name: "${esp_name} Up"
    internal: true  

  - platform: gpio
    id: switch_down
    pin: 26
    name: "${esp_name} Down"
    internal: true
 
binary_sensor:
  - platform: gpio
    id: ext_button_up
    pin: 
      number: 27
      mode:
        input: true
        pulldown: true 
    name: "${esp_name} Ext Up"
    on_press:
      then:
        - switch.turn_on: switch_up
    on_release:
      then:
        - switch.turn_off: switch_up

  - platform: gpio
    id: ext_button_down
    pin: 
      number: 14
      mode:
        input: true
        pulldown: true 
    name: "${esp_name} Ext Down"
    on_press:
      then:
        - switch.turn_on: switch_down
    on_release:
      then:
        - switch.turn_off: switch_down

script:
  - id: single_down
    mode: queued
    then:
      - switch.turn_on: switch_down
      - delay: ${push_delay}
      - switch.turn_off: switch_down
      - delay: ${push_delay}      

  - id: single_up
    mode: queued
    then:
      - switch.turn_on: switch_up
      - delay: ${push_delay}
      - switch.turn_off: switch_up
      - delay: ${push_delay}    

  - id: stop
    then:
      - if:
          condition:
            lambda: |-
              if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_IDLE) {
                return true;
              } else if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_OPENING) {
                return false;
              } else if (id(tb_cover).current_operation == CoverOperation::COVER_OPERATION_CLOSING) {
                return false;
              }            
          then:
            - button.press: button_double_down
          else:
            - button.press: button_single_down
 
button:
  - platform: template
    name: "${esp_name} Single Down" 
    id: button_single_down
    on_press:
      - script.execute: single_down

  - platform: template
    name: "${esp_name} Double Down" 
    id: button_double_down
    on_press:
      - repeat:
          count: 2
          then:
            - script.execute: single_down

  - platform: template
    name: "${esp_name} Triple Down" 
    id: button_triple_down
    on_press:
      - repeat:
          count: 3
          then:
            - script.execute: single_down

  - platform: template
    name: "${esp_name} Quad Down" 
    id: button_quad_down
    on_press:
      - repeat:
          count: 4
          then:
            - script.execute: single_down   

  - platform: template
    name: "${esp_name} Single Up" 
    id: button_single_up
    on_press:
      - script.execute: single_up

  - platform: template
    name: "${esp_name} Double Up" 
    id: button_double_up
    on_press:
      - repeat:
          count: 2
          then:
            - script.execute: single_up  

  
cover:
  - platform: time_based
    id: tb_cover
    name: "LR Side Cover"
    has_built_in_endstop: true
    open_duration: 20s    
    close_duration: 10s

    open_action:
      - button.press: button_single_up

    close_action:
      - button.press: button_single_down

    stop_action:
      - script.execute: stop
```

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by xprofiler**
# What does this implement/fix?

I need to know the last operation of the cover when handle a cover STOP command, because the current_operation is already set to IDLE. With the last operation information I am able to decide which of the two buttons (UP and DOWN) for my Warema controller I have to press to stop the cover.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/4252

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
cover:
  - platform: time_based
    name: "Sunblind AZ"
    id: sunblind_az
    open_action:
      - button.press: az_h
    open_duration: 34s
    close_action:
      - button.press: az_r
    close_duration: 32.5s
    stop_action: 
      - lambda: !lambda |-
          if (id(sunblind_az).last_operation() == CoverOperation::COVER_OPERATION_IDLE) {
            // Cover is idle
          } else if (id(sunblind_az).last_operation() == CoverOperation::COVER_OPERATION_OPENING) {
            // Cover is currently opening
            id(az_r).press();
          } else if (id(sunblind_az).last_operation() == CoverOperation::COVER_OPERATION_CLOSING) {
            // Cover is currently closing
            id(az_h).press();
          }
    has_built_in_endstop: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). See https://github.com/esphome/esphome-docs/pull/3532

### CODE REVIEW

**Comment by xprofiler** [comment]
I created my feature branch from tag https://github.com/esphome/esphome/tree/2023.12.5 because this is my docker containers version and the version I tested my changes.

**Review by kroimon — CHANGES_REQUESTED** [review]
Please rebase your branch onto the latest dev branch, there are a lot of unrelated changes in this PR.

**Comment by xprofiler** [comment]
For some reason a normal rebase was not possible, so I recreated my feature branch derived from the current dev branch and make this one line change to corresponding file. I will try to test it with a current build.

File: esphome/components/time_based/time_based_cover.h [hunk 1]
```diff
@@ -23,6 +23,7 @@ class TimeBasedCover : public cover::Cover, public Component {
   void set_has_built_in_endstop(bool value) { this->has_built_in_endstop_ = value; }
   void set_manual_control(bool value) { this->manual_control_ = value; }
   void set_assumed_state(bool value) { this->assumed_state_ = value; }
+  cover::CoverOperation last_operation() const { return this->last_operation_; }
```
Comment (jesserockz) [inline esphome/components/time_based/time_based_cover.h]:
```suggestion
  cover::CoverOperation get_last_operation() const { return this->last_operation_; }
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**